src/gc-arena/src/types.rs

```diff
@@ -48,10 +61,41 @@ impl GcFlags {
         self.0.get() & 0x4 != 0x0
     }
 
+    /// This is `true` if we've traced a weak pointer during to this `GcBox`
+    /// during the most recent `Phase::Propagate`. This is reset back to
+    /// `false` during `Phase::Sweep`.
+    #[inline]
+    pub(crate) fn has_weak_ref(&self) -> bool {
+        self.0.get() & 0x8 != 0x0
+    }
+
+    /// Determines whether or not we've dropped the stored `dyn Collect` value.
+    /// When we garbage-collect a `GcBox` that still has outstanding weak pointers,
+    /// we set `alive` to false. When there are no more weak pointers remaining,
+    /// we will deallocate the `GcBox`, but skip dropping the `dyn Collect` value
+    /// (since we've already done it).
+    #[inline]
+    pub(crate) fn alive(&self) -> bool {
+        self.0.get() & 0x10 != 0x0
+    }
+
+    #[inline]
     pub(crate) fn set_needs_trace(&self, needs_trace: bool) {
         self.0
             .set((self.0.get() & !0x4) | if needs_trace { 0x4 } else { 0x0 });
     }
+
+    #[inline]
+    pub(crate) fn set_has_weak_ref(&self, has_weak_ref: bool) {
+        self.0
+            .set((self.0.get() & !0x8) | if has_weak_ref { 0x8 } else { 0x0 });
+    }
+
+    #[inline]
+    pub(crate) fn set_alive(&self, alive: bool) {
```

Comment:
again, `set_live`, or `set_dropped`... or actually don't even have a method to set this flag and instead just have a method to drop the value early (also store the value in a `ManuallyDrop`)

Reply:
I've renamed this to `set_live`. However, we still need a flag when we're using `ManuallyDrop`, since we need to track whether or not we've called `ManuallyDrop::drop`.

Reply:
I realized that, i was thinking more like instead of setting the is_live flag, you would have a method that 1) asserts that `is_live` is true, 2) sets it to false, 3) drops the type, and then in the destructor, check the flag and drop if it's not set.  Just as a way of removing one way of getting it wrong, but it's not actually a super big deal really.